File: server/base_consumer/models/workflow.go
package models

import "gorm.io/gorm"

type EventType string

const (
	ActionEventType   EventType = "action"
	ReactionEventType EventType = "reaction"
)

type Parameters struct {
	gorm.Model
	Name        string `json:"name"`
	Description string `json:"description"`
	Type        string `json:"type"`
	EventID     uint   `gorm:"foreignKey:EventID" json:"event_id"`
}

type Workflow struct {
	gorm.Model
	UserID      uint           `gorm:"foreignKey:UserID" json:"user_id"`
	Name        string         `json:"name"`
	Description string         `json:"description"`
	Status      WorkflowStatus `gorm:"type:enum('pending', 'processed', 'failed')" json:"status"`
	IsActive    bool           `json:"is_active"`
	Events      []Event        `gorm:"many2many:workflow_events" json:"events"`
}

func (w *Workflow) BeforeCreate(tx *gorm.DB) (err error) {
	if w.Status == "" {
		w.Status = WorkflowStatusPending
	}
	w.IsActive = true
	return
}

File: client_mobile/lib/data/workflow.dart
import 'package:area/data/action.dart';

class WorkflowEvent {
  final WorkflowActionReaction action;
  final String type;

  WorkflowEvent({required this.action, required this.type});
}

class Workflow {
  final String name;
  final String description;
  final List<int> servicesId;
  final List<WorkflowEvent> events;
  final String? status;
  int id;

  Workflow({
    required this.name,
    required this.description,
    required this.servicesId,
    required this.events,
    this.status,
    this.id = 0
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'services': servicesId,
      'events': events.map((event) => event.action.toJson(event.type)).toList(),
    };
  }

   factory Workflow.fromJson(Map<String, dynamic> json) {
    try {
      return Workflow(
        id: json['ID'] ?? 0,
        name: json['name'] ??
            'Unknown',
        description: json['description'],
        status: json['status'],
        servicesId: [],
        events: []
      );
    } catch (e) {
      print('Error parsing Workflow: $e');
      return Workflow(
          name: 'Unknown', status: "failed", description: "Unknown", servicesId: [], events: []);
    }
  }
}
